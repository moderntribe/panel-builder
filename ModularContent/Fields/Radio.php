<?php


namespace ModularContent\Fields;

use ModularContent\Panel;


/**
 * Class Radio
 *
 * @package ModularContent\Fields
 *
 * A group of radio buttons.
 */
class Radio extends Select {

	protected $layout       = 'vertical';
	protected $option_width = 12;

	public function __construct( $args = [] ) {
		$this->check_layout( $args );
		$this->check_option_width( $args );

		$this->defaults['layout']       = $this->layout;
		$this->defaults['option_width'] = $this->option_width;
		parent::__construct( $args );
	}

	public function get_blueprint() {
		$blueprint                 = parent::get_blueprint();
		$blueprint['layout']       = $this->layout;
		$blueprint['option_width'] = $this->option_width;

		return $blueprint;
	}

	protected function check_layout( $args ) {
		if ( isset( $args['layout'] ) && $args['layout'] !== 'vertical' && $args['layout'] !== 'horizontal' && $args['layout'] !== 'inline' ) {
			throw new \LogicException( 'Layout argument can only be "vertical", "horizontal", or "inline".' );
		}
	}

	protected function check_option_width( $args ) {
		if ( isset( $args['option_width'] ) && ( $args['option_width'] < 1 || $args['option_width'] > 12 ) ) {
			throw new \LogicException( 'option_width argument must be between 1-12.' );
		}
	}

}