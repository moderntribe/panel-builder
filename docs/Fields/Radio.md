## Radio

A group of radio inputs.

### Parameters

||||
|---|---|---|
| `label`       | `string` | The __() translated label for this field       |
| `name`        | `string` | The unique name for this field                 |
| `description` | `string` | The __() translated description for this field |
| `options` | `array` | An array of key => value pairs in which the `key` is the option name and the `value` is the option Label |
| `layout` | `string` | The type of layout to use; either `vertical`, `horizontal`, or `inline` |
| `option_width` | `int` | The width for this input if it is part of a group. Between 1-12 |
### Example

```php
$field = new Radio( [
    'label' => __('Pick One'),
    'name' => 'my-field',
    'description' => __( 'Pick the thing that you pick' )
    'layout' => 'vertical',
    'option_width' => 6,
    'options' => [
        'first' => __( 'The First Option' ),
        'second' => __( 'The Second Option' ),
    ]
] );
```