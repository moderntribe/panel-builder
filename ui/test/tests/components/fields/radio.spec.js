/* global expect b:true */

import React from 'react';
import { shallow, mount } from 'enzyme';
import Radio from 'components/fields/radio';
import renderer from 'react-test-renderer';

// test data
const OPTIONS1 = [
	{
		label: 'Standard',
		value: 'standard',
	},
	{
		label: 'Cards',
		value: 'cards',
	},
	{
		label: 'Full',
		value: 'full',
	},
];

const OPTIONS2 = [
	{
		label: 'Orange',
		value: '1',
	},
	{
		label: 'Raspberry',
		value: '2',
	},
];


describe('Radio field', () => {
	it('renders correctly', () => {
		const tree = renderer
			.create(<Radio description="This is a test description" label="Some test label" options={OPTIONS1} />)
			.toJSON();
		expect(tree).toMatchSnapshot();
	});

	it('renders as a <div>', () => {
		const wrapper = shallow(<Radio />);
		expect(wrapper.type()).toEqual('div');
	});

	it('has no radio buttons without properties.', () => {
		const wrapper = shallow(<Radio />);
		expect(wrapper.find('input').length).toEqual(0);
	});

	it('has a label', () => {
		const wrapper = shallow(<Radio />);
		expect(wrapper.find('label').length).toEqual(1);
	});

	it('allows us to set and edit props', () => {
		const wrapper = mount(<Radio description="This is a test description" label="Some test label" options={OPTIONS1} />);
		expect(wrapper.props().description).toEqual('This is a test description');
		wrapper.setProps({ description: 'Change description for test' });
		expect(wrapper.props().description).toEqual('Change description for test');

		expect(wrapper.props().label).toEqual('Some test label');
		wrapper.setProps({ label: 'Change label for test' });
		expect(wrapper.props().label).toEqual('Change label for test');

		expect(wrapper.props().options).toEqual(OPTIONS1);
		wrapper.setProps({ options: OPTIONS2 });
		expect(wrapper.props().options).toEqual(OPTIONS2);
	});

	it('allows us to properly formatted radio options', () => {
		const wrapper = mount(<Radio description="This is a test description" label="Some test label" options={OPTIONS1} />);
		expect(wrapper.find('label.plradio-label').length).toEqual(3);
		expect(wrapper.find('label.plradio-label').first().find('input').length).toEqual(1);
		expect(wrapper.find('label.plradio-label').first().find('input').props().value).toEqual('standard');
		expect(wrapper.find('label.plradio-label').first().text()).toEqual('Standard');
	});
});
