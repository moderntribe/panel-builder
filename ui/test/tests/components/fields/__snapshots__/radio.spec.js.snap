exports[`Radio field renders correctly 1`] = `
<div
  className="field vertical panel-field panel-conditional-field">
  <label
    className="label panel-field-label">
    Some test label
    <span
      className="icon"
      onBlur={undefined}
      onClick={undefined}
      onContextMenu={undefined}
      onFocus={undefined}
      onMouseDown={undefined}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      onTouchStart={undefined} />
  </label>
  <ul
    className="list">
    <li
      className="item"
      style={Object {}}>
      <label
        className="radioLabel plradio-label">
        <input
          checked={false}
          className="radio"
          data-depth={0}
          data-field="radio"
          data-option-type="single"
          name="modular-content-"
          onChange={[Function]}
          tabIndex={0}
          type="radio"
          value="standard" />
        <span />
        Standard
      </label>
    </li>
    <li
      className="item"
      style={Object {}}>
      <label
        className="radioLabel plradio-label">
        <input
          checked={false}
          className="radio"
          data-depth={0}
          data-field="radio"
          data-option-type="single"
          name="modular-content-"
          onChange={[Function]}
          tabIndex={0}
          type="radio"
          value="cards" />
        <span />
        Cards
      </label>
    </li>
    <li
      className="item"
      style={Object {}}>
      <label
        className="radioLabel plradio-label">
        <input
          checked={false}
          className="radio"
          data-depth={0}
          data-field="radio"
          data-option-type="single"
          name="modular-content-"
          onChange={[Function]}
          tabIndex={0}
          type="radio"
          value="full" />
        <span />
        Full
      </label>
    </li>
  </ul>
</div>
`;
